keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,494 @@
+
+Introducing this schema will include three things: 
+* Schema definition format
+* Set of helper functions to populate this schema (e.g. `WithObject` for referencing k8s object)
+* Validation mechanism ensuring schema is not bypassed by lower level logging API
+
+TODO(serathius): Propose how definition format should look like, probably yaml similar to https://github.com/elastic/ecs/tree/master/schemas
+
+### Migration
+ 
+Migration to new structured logging API will be divided into 4 stages:
+
+#### Stage 1 / Alpha
+
+This stage needs to be performed in a single release of Kubernetes. When finished for every **existing** log entry produced in legacy output format there should be a corresponding log entry in json/dev format containing similar information possibly with some additional context information attached to it.
+When using the legacy log format after this stage Kubernetes control plane components should produce exactly the same logs as before this stage. 
+After this stage format and content of json/dev formatted logs will be still in pre-alpha state and most likely will change after manual review. 
+After completing this stage adding **new log entries should require using only the new logr-based logging API**. Double writing using both klog and logr APIs for new log entries should not be needed.  
+
+#### Preparation
+* Develop and merge to Kubernetes master 3 implementations/configurations of logr API which will be needed in the latter steps of the migration:
+    * `no-op` implementation which does not log anything,
+    * `json` log format implementation,
+    * `dev` log format implementation.
+* Implement `k8slog` package which during its initialization should read `--logging-format` flag and initialize two global variables providing access to:
+    * logr implementation to be used to migrate existing logs (json/dev loggers for json/dev log format and no-op logger for legacy format)
+    * logr implementation to be used by new logs (json/dev loggers for json/dev log format and klogr logger for legacy format)
+* For each package in Kubernetes source code define a logging module to which it will belong. Each logging module can span across multiple packages.
+* For each logging module define two variables with logr implementations to be used for existing and new logs initialized with the name of the module. Those should only be used until stage 2 of the migration is fully completed.
+* Implement a package implementing klog API which would check a global variable and either passthrough call to klog (with a proper deep value handing) or do nothing. This global variable should be initialized based on `--logging-format` value.
+* Kubernetes and all of its dependencies should be upgraded to klog v2.
```

Comment:
"klog v2" is used here for the first time, what exactly does this refer to?

Reply:
By "klog v2" I meant "k8s.io/klog/v2" which is current master of klog. If for some reason switching all dependencies to use new klog API will not be possible we can create our own wrapper for klog which will redirect logs to logr when log output is set to json and enforce its usage by using gomod replace directive.